google/cloud/aiplatform/docker_utils/build.py

```diff
@@ -0,0 +1,431 @@
+            Optional. A comment explaining the copy operation.
+    """
+    cmd = "COPY {}\n".format(json.dumps([from_path, to_path]))
+
+    if comment is not None:
+        formatted_comment = "\n# ".join(comment.split("\n"))
+        return textwrap.dedent(
+            """
+            # {}
+            {}
+            """.format(
+                formatted_comment, cmd,
+            )
+        )
+
+    return cmd
+
+
+def _prepare_dependency_entries(
+    setup_path: Optional[str] = None,
+    requirements_path: Optional[str] = None,
+    extra_packages: Optional[List[str]] = None,
+    extra_requirements: Optional[List[str]] = None,
+    extra_dirs: Optional[List[str]] = None,
+    pip_command: str = "pip",
+):
+    """Returns the Dockerfile entries required to install dependencies.
+
+    Args:
+        setup_path (str):
+            Optional. The path that points to a setup.py.
+        requirements_path (str):
+            Optional. The path that points to a requirements.txt file.
+        extra_packages (List[str]):
+            Optional. The list of user custom dependency packages to install.
+        extra_requirements (List[str]):
+            Optional. The list of required dependencies to be installed from remote resource archives.
+        extra_dirs (List[str]):
+            Optional. The directories other than the work_dir required.
+        pip_command (str):
+            Required. The pip command used for install packages.
+    """
+    ret = ""
+
+    if setup_path is not None:
+        ret += _generate_copy_command(
+            setup_path,
+            "./setup.py",
+            comment="setup.py file specified, thus copy it to the docker container.",
+        ) + textwrap.dedent(
+            """
+            RUN {} install --no-cache-dir .
+            """.format(
+                pip_command,
+            )
+        )
+
+    if requirements_path is not None:
+        ret += _generate_copy_command(
+            requirements_path,
+            "./requirements.txt",
+            comment="requirements.txt file specified, thus copy it to the docker container.",
+        ) + textwrap.dedent(
+            """
+            RUN {} install --no-cache-dir -r ./requirements.txt
+            """.format(
+                pip_command,
+            )
+        )
+
+    if extra_packages is not None:
+        for extra in extra_packages:
+            package_name = os.path.basename(extra)
+            ret += textwrap.dedent(
+                """
+                {}
+                RUN {} install --no-cache-dir {}
+                """.format(
+                    _generate_copy_command(extra, package_name),
+                    pip_command,
+                    shlex_quote(package_name),
+                )
+            )
+
+    if extra_requirements is not None:
+        for requirement in extra_requirements:
+            ret += textwrap.dedent(
+                """
+                RUN {} install --no-cache-dir --upgrade {}
```

Comment:
Why is --upgrade only specified for extra_requirements?

Reply:
Removed it since it's not necessary.